Lib/test/test_zoneinfo/_support.py

```diff
@@ -0,0 +1,54 @@
+import contextlib
+import functools
+import sys
+import threading
+import unittest
+from test.support import import_fresh_module
+
+OS_ENV_LOCK = threading.Lock()
+TZPATH_LOCK = threading.Lock()
+TZPATH_TEST_LOCK = threading.Lock()
+
+
+@functools.lru_cache(1)
+def get_modules():
+    import zoneinfo as c_module
+    py_module = import_fresh_module("zoneinfo", blocked=["_czoneinfo"])
+
+    return py_module, c_module
```

Comment:
It's not immediately obvious to me what the purpose of `functools.lru_cache(1)` is in this context. Is that to ensure the module only gets imported once? Does it potentially get imported multiple times otherwise and cause issues?

Regardless of the reason though, I think a brief comment to explain might be helpful.

Reply:
This sort of thing is very touchy and can cause some headscratching bugs because it's trying to manipulate the import system in ways it wasn't really intended to be manipulated. For example, the tests will stop passing if you switch the order of those two import statements for reasons that I can guess at but which still aren't *entirely* clear.

In any case, it's safe to say that this is not idempotent and if I were being very belt-and-suspenders I'd also acquire a lock as well (hey hey, here's one of those use cases @rhettinger was asking for where a `@call_once` decorator might need the property that the function is actually only executed exactly once), but we can save that particular complication for a (more likely than you might think) future where there's some possibility that two threads will attempt to call this function at the same time.

At the moment I don't think it's very important because it's only used in `test_zoneinfo.py`, but that's only because I didn't migrate [the property tests](https://github.com/pganssle/zoneinfo/blob/ffd21a6d065e04725e04b37bb430c2559fefd5fa/tests/test_zoneinfo_property.py) from the reference implementation. My hope is that we'll eventually integrate property testing into the standard library.

Anyway, that's a long explanation. I'll try to distill it into a few lines and add a docstring tomorrow. Thanks!

Reply:
No problem, I appreciate the detailed explanation. And indeed, this does seem like it would make for a good use case of the proposed `@call_once` from the recent discussion in python-ideas and python-dev.

> My hope is that we'll eventually integrate property testing into the standard library.

I hope so as well, although admittedly it's not something that I have much personal experience with. I would certainly see testing a much wider range of inputs as a significant plus.

Reply:
OK, I refactored this a bit, and also switched to a `call_once` decorator that is thread-safe, because why not.